### PR TITLE
Add spec.api.onlyBindToAddress configuration

### LINKED ISSUE #957: Changing "bind-address" of kube-api
I have systems with multiple network interfaces and I would like to initialize the cluster on only one of the interfaces, say 192.168.1.1.

## TLDR
Adding `spec.api.extraArgs.bind-address = 192.168.1.1` (and adding to sans) in k0s config successfully causes the kube-api server to only bind on 192.168.1.1. However, when changing the `/var/lib/k0s/pki/admin.conf`s `clusters[0].cluster.server = https://192.168.1.1:6443`, `k0s token create --role=worker` results in a `TLS handshake timeout`. 

## The rest
This issue is more a question of how to configure k0s to achieve what I am looking to do. I don't want the kube-api server listening on every interface in this case, but changing the bind-address breaks being able to access the cluster. I know it is a certificate issue, I'm just not sure where.

Here is the relevant section in my k0s.yaml file:
```yaml
spec:
  api:
    address: 192.168.1.1
    port: 6443
    k0sApiPort: 9443
    externalAddress: 192.168.1.1
    extraArgs:
      bind-address: 192.168.1.1
    sans:
    - 192.168.1.1
```

For context, this is a single node control plane cluster that manages airgapped nodes that are behind a specific network interface.

### PR DESCRIPTION
**Pull request by pschichtel**
I picked up #1038 and rebased it onto main. Smoke tests pass, but not entirely clear what was left here.

From #1038:

> Should a new `bindAddress` option be created and change documentation for the existing `address` option, or should the existing `address` option's behavior be changed to match the existing documentation?
> 
> The later I can see causing some issues with existing setups, `bindAddress` is probably the safest route. Currently the default for `bindAddress` is "0.0.0.0", which is the default kube-api-server uses if the `--bind-address` flag isn't used.
> 
> **Issue** Fixes #957
> 
> **What this PR Includes** Adds the ability to define a `bindAddress` for the kube-api-server.

### CODE REVIEW

File: pkg/apis/k0s/v1beta1/api.go
Comment (twz123) [reply pkg/apis/k0s/v1beta1/api.go:115]: OTOH, if this would fall back to `localhost` if BindAddess is empty, it would make some sense again, and it would also mean that there'd be no behavioral change at all when BindAddress is not specified.

File: pkg/apis/k0s/v1beta1/api.go
Comment (twz123) [inline pkg/apis/k0s/v1beta1/api.go:39]: We might want to add some kubebuilder validation here to reject nonsensical values already at the API server level. Sure, a regex that's correctly checking for IPv4/IPv6 compliance will probably be one megabyte in size or so, but something short that rejects obviously wrong stuff might be helpful nevertheless?

**Review by twz123 — COMMENTED** [review]
Thanks for picking this up.

I'd double-check if there are some bad interactions with node-local load balancing? I imagine the LB config needs to pick up the correct bind addresses as well?

We should think about some integration test, too, I suppose.

File: pkg/apis/k0s/v1beta1/api.go
Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go:39]: what are you thinking of? something like `^([0-9.]+|[0-9a-f:]+)$` would at least ensure that the correct alphabet is being used while keeping it simple.

File: docs/configuration.md
Comment (pschichtel) [reply docs/configuration.md:58]: I guess there is no equivalent to example.org in ip addresses, maybe some private IP address? 192.168.0.1 mabye ?

Comment (pschichtel) [reply docs/configuration.md:58]: nvm, there actually is: https://datatracker.ietf.org/doc/html/rfc5737

I'm going for TEST-NET-1's 192.0.2.1

**Comment by twz123** [comment]
Okay, some things to proceed with:

* We need an integration test. Need to come up with a scenario for that, i.e. what is to be tested.
* We need to double-check if this plays nicely with nllb and other things. Do we need to update the docs on e.g. external load balancing? How would this affect k0sctl, if at all?

File: pkg/apis/k0s/v1beta1/api.go
Comment (twz123) [reply pkg/apis/k0s/v1beta1/api.go:40]: Yeah, something like this. Although I have a gut feeling that the current regex will break IPv6 link-local addresses. Maybe my idea wasn't the best here :smile:. A pity that there's no "ip" format in the OpenAPI validator, just "ipv4" _or_ "ipv6" (at least as far as I reckon). A workaround would be to have `bindAddress: {type: IPv4, ipv4: 127.0.0.1}` and `bindAddress: {type: IPv6, ipv6: "::1"}`, just to have two distinct fields to put the format tags on. That'd be a _very_ artificial distinction, though.

Overall, we might be better off without server-side validation, unfortunately, just to not introduce unnecessary bugs.

Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go:40]: > Although I have a gut feeling that the current regex will break IPv6 link-local addresses.

that's why I added the `%` in a later revision, but I'm still not familiar enough with the related RFCs to be certain it covers everything.

I also lean towards not risking it.

**Comment by pschichtel** [comment]
> We need an integration test. Need to come up with a scenario for that, i.e. what is to be tested.

I'm not very familiar with the code base yet, can you point me to examples for integration tests?

> We need to double-check if this plays nicely with nllb and other things.

I'd have to look into how nllb works exactly. Depending on how the upstream connection is made its configuration probably needs adjustments.

> Do we need to update the docs on e.g. external load balancing?

I guess a disclaimer like "if you change the bind address make sure to align your external load balancer" or so. I feel like people changing bind addresses are probably aware of the impact. But that's just my opinion.

> How would this affect k0sctl, if at all?

k0sctl would need to interact with the API server on the correct IP, wouldn't it?

**Comment by twz123** [comment]
Sorry for getting back to you so late @pschichtel. I'll bluntly blame the holiday season for that. :sweat_smile:

> > We need an integration test. Need to come up with a scenario for that, i.e. what is to be tested.
> 
> I'm not very familiar with the code base yet, can you point me to examples for integration tests?
> 
> > We need to double-check if this plays nicely with nllb and other things.
> 
> I'd have to look into how nllb works exactly. Depending on how the upstream connection is made its configuration probably needs adjustments.

We could try to start with the [existing nllb inttest](https://github.com/k0sproject/k0s/tree/e121de6a33b864bc3db74c8a23a9d488f9538177/inttest/nllb), copy that over into a separate directory and start hacking on the new one. You can probably delete all the cluster-restart stuff in there, and just rely on the cluster bootstrapping part. You need to add the new directory name to [`inttest/Makefile.variables`](https://github.com/k0sproject/k0s/tree/e121de6a33b864bc3db74c8a23a9d488f9538177/inttest/Makefile.variables). Then you should be able to do something like `make check-yolo` (or whatever the name is you choose.)

Might make sense to simply try to overwrite the bind addresses of every component that you can get hold of and see what happens to the cluster.

> > Do we need to update the docs on e.g. external load balancing?
> 
> I guess a disclaimer like "if you change the bind address make sure to align your external load balancer" or so. I feel like people changing bind addresses are probably aware of the impact. But that's just my opinion.

Sounds good enough for the start.

> 
> > How would this affect k0sctl, if at all?
> 
> k0sctl would need to interact with the API server on the correct IP, wouldn't it?

Yes. Would it need to look at the k0s config to figure this out? Or would it "just work"? Maybe @kke can tell.

**Comment by pschichtel** [comment]
> Sorry for getting back to you so late @pschichtel. I'll bluntly blame the holiday season for that. 😅
 
No worries, I don't have any rush with this one.

> We could try to start with the [existing nllb inttest](https://github.com/k0sproject/k0s/tree/e121de6a33b864bc3db74c8a23a9d488f9538177/inttest/nllb), copy that over into a separate directory and start hacking on the new one. You can probably delete all the cluster-restart stuff in there, and just rely on the cluster bootstrapping part. You need to add the new directory name to [inttest/Makefile.variables](https://github.com/k0sproject/k0s/tree/e121de6a33b864bc3db74c8a23a9d488f9538177/inttest/Makefile.variables). Then you should be able to do something like make check-yolo (or whatever the name is you choose.)
> 
> Might make sense to simply try to overwrite the bind addresses of every component that you can get hold of and see what happens to the cluster.

Makes sense, I'll have a go at that.

**Comment by twz123** [comment]
(The linter error will go away after a rebase. See #3991 for details.)

**Comment by pschichtel** [comment]
yeah I plan to work on this again towards the end of this week

**Comment by andycandy-de** [comment]
Any idea when this feature will be available? I also have multiple IPs and only want to bind the k0scontroller to a specific IP.

**Comment by pschichtel** [comment]
@andycandy-de sadly this had to be pushed back a little in my prio list, but I still intent to eventuelly finish this up. I think some of the recent k0s/k8s releases also added additional components, which probably needs/deserves some investigation. I'm currently unable to work on it until mid of April, so no sooner than that.

File: pkg/apis/k0s/v1beta1/api.go
Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go:39]: Is this something that could be deferred to after this PR ?

Comment (twz123) [reply pkg/apis/k0s/v1beta1/api.go:39]: Yeah, let's just skip validation right now, and just :crossed_fingers:

File: pkg/apis/k0s/v1beta1/api.go
Comment (twz123) [inline pkg/apis/k0s/v1beta1/api.go:65]: Any reason for defaulting this to "0.0.0.0"?

File: pkg/apis/k0s/v1beta1/api_test.go
Comment (twz123) [inline pkg/apis/k0s/v1beta1/api_test.go:39]: Do we need to add the BindAddress to theexisting test cases?

File: pkg/apis/k0s/v1beta1/api_test.go
Comment (twz123) [inline pkg/apis/k0s/v1beta1/api_test.go:78]: :stuck_out_tongue:

```suggestion
			BindAddress: "something.that.is.not.valid//(())",
```

File: pkg/apis/k0s/v1beta1/api.go
Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go:65]: Compared to just leaving it out?

File: pkg/apis/k0s/v1beta1/api_test.go
Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api_test.go:39]: As in: should we leave them untouched?

File: pkg/apis/k0s/v1beta1/api.go
Comment (twz123) [reply pkg/apis/k0s/v1beta1/api.go:65]: Yes.

File: pkg/apis/k0s/v1beta1/api.go
Comment (twz123) [inline pkg/apis/k0s/v1beta1/api.go:36]: Now I feel really stupid for only realizing it now, but this parameter actually sounds like what we want to do here.

We use this parameter as the advertise address, not as the bind address. Would the following assumption hold?

"The advertise address will work when connecting to the API server locally, as well as from other peer controllers."

If yes, then the implementation could be much easier: Just fix all the places where k0s uses a hard coded localhost or 127.0.0.1 to connect to the API server by using the `Address`. This is pretty much all the places where `BindAddress` is used now. Then set `spec.api` to sth. like `{"address": "x.x.x.x", "extraArgs": {"bind-address": "x.x.x.x"}}`. And we're done. (At least that's what I'd guess :see_no_evil:)

**Comment by twz123** [comment]
>But how would I go about that in the current containerized setup, given that I have limited control over what networks are being used.

Might be sufficient if we bind to something else than 0.0.0.0 or a loopback address. So what you could try to do is to bind to the Docker container's IP. In the integration test, you could get that via

```go
controller0, err := s.MachineForName(s.ControllerNode(0))
s.Require().NoError(err)
ip := controller0.Status().IP
```

File: pkg/apis/k0s/v1beta1/api.go
Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go:36]: I noticed address myself last time I worked on this, totally forgot to add a comment about that :D I'll look into it.

Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go]: @twz123 the latest commit replaces BindAddress with Address as suggested.

Both versions (with dedicated BindAddress and with Address used for bind address) seem to work.

@gakio12 in the original PR #1038 actually asked the question:

> Should a new bindAddress option be created and change documentation for the existing address option, or should the existing address option's behavior be changed to match the existing documentation?

To which @jnummelin answered, that he prefers the dedicated option.

How would we justify (and document) the existence of both Address BindAddress? I see one possible reason to separate them: `0.0.0.0` as a bind address ([which also came up in the previous PR.](https://github.com/k0sproject/k0s/pull/1038#pullrequestreview-817786306)). If address is something I want to use to connect to the API server, then it should never be `0.0.0.0`. If address is something the server binds to, then I'd expect to be able to set `0.0.0.0`.

Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go]: The broken tests seem to be broken by the switch over to Address also being the bind address and I think that makes sense. I just attempted to change it so Address takes to role of BindAddress and merge its previous role with ExternalAddress, but that falls apart quickly.

My current understanding is:

* Address: Used internally by the cluster
* ExternalAddress: Used by clients external to the cluster
* BindAddress: The address the api server binds to

Clearly from this, if BindAddress is set, than Address must be compatible with its value, because we can't bind the API server to e.g. 10.0.0.1 and then advertise it's address as 10.0.0.2. But we can bind it to 0.0.0.0 and then advertise it as any address.

New proposal for the options:

* Address: IP used internally by the cluster and used for the `--bind-address` option in API server
* ExternalAddress: Address used by clients external to the cluster (does this one actually need to be an IP address?)
* OnlyBindToAddress: Boolean option (default false for compatibility) to allow binding the API server to only the IP address given by Address instead of 0.0.0.0/::

Address and ExternalAddress would also be validated to not be 0.0.0.0 or ::.

**Comment by pschichtel** [comment]
I'll try to have a look at the openebs test this evening

**Comment by pschichtel** [comment]
The test actually passed on my machine

**Comment by pschichtel** [comment]
@twz123 @jnummelin given this is now the oldest open PR and I haven't heard back in a while: Is there still interest in this? I've been keeping it in sync with the master so far. I myself don't have a need for this anymore, but I'm still willing to push it over the finish line.

**Comment by andycandy-de** [comment]
@pschichtel I am still waiting for this feature to get merged... Thank you for your efforts... ❤️

@twz123 @jnummelin any chance to get the feature merged soon?

**Comment by jr200** [comment]
I'm also interested in this PR being merged. Thanks all.

**Comment by pschichtel** [comment]
@jr200  @andycandy-de has any one if you guys given this PR a spin in a lab environment? It's fairly simple to use custom builds with k0s

**Comment by jr200** [comment]
> @jr200 @andycandy-de has any one if you guys given this PR a spin in a lab environment? It's fairly simple to use custom builds with k0s. It would also be pretty easy to backport it to older stable releases.

Hi @pschichtel,

I had a go just now and I hit some dead-ends. I'm using `k0sctl` on macos (intel). I checked out your branch, installed docker-desktop, did `make`, which I then realised built the linux version of the binaries. I tried a manual `go` build from your folder which failed without producing any binaries on the error `pkg/component/worker/containerd/component.go:121:9: undefined: winExecute`. 

Some CIs for open-source projects are setup to automatically build binaries on all platforms which can be tested by downloading them from the CI directly. Can you generate a macos binary binary somehow - i'd be happy to test...

By the way, I think this PR only partially solves my issue. I'm using k0s over zerotier VPN, so I wanted all k0s-related ports to bind on the IP from a specific network-interface (i.e., the VPN's NIC). This patch doesn't seem to set `--advertise-address` on the `spec.api.extraArgs` - should it?

**Comment by pschichtel** [comment]
> I'm using k0sctl on macos (intel). I checked out your branch, installed docker-desktop, did make, which I then realised built the linux version of the binaries. I tried a manual go build from your folder which failed without producing any binaries on the error pkg/component/worker/containerd/component.go:121:9: undefined: winExecute.

I don't have the capacity to help you with macos build issues, sorry.

> This patch doesn't seem to set --advertise-address on the spec.api.extraArgs - should it?

Isn't this where it's set: https://github.com/k0sproject/k0s/pull/3824/files#diff-e2c56849a04d7e414d5b968597251ab14c9f4229772cca9babf39600f63d3991R102 ? That would be configured by the "address" option k0s already has.

**Comment by pschichtel** [comment]
it's been months of rebasing now.

File: docs/configuration.md
Comment (twz123) [inline docs/configuration.md:120]: Would be great to add some details as to why this knob exists, and what may be the consequences if set to true. Basically some excerpt of https://github.com/k0sproject/k0s/pull/3824#discussion_r1622862280.

File: pkg/component/controller/apiserver.go
Comment (twz123) [inline pkg/component/controller/apiserver.go:237]: Ideally, this would use `net.JoinHostPort`. I think it'll break for IPv6 addresses otherwise.

File: cmd/controller/certificates.go
Comment (twz123) [inline cmd/controller/certificates.go:67]: Ideally, this would use `net.JoinHostPort`. I think it'll break for IPv6 addresses otherwise.

File: docs/configuration.md
Comment (twz123) [inline docs/configuration.md:58]: This is a remnant, right?

File: docs/configuration.md
Comment (twz123) [inline docs/configuration.md:55]: As it defaults to `false` and it's `omitempty`, the `k0s kubeconfig create` command won't include that field.

```suggestion
```

File: docs/configuration.md
Comment (twz123) [inline docs/configuration.md:120]: I wonder if we could just keep `bindAddress` instead of `onlyBindToAddress` and let the user decide which address to bind to instead of forcing it to be either the advertised address or all interfaces. This may be a niche requirement, though. I cannot really judge. Maybe some folks with more network expertise have an opinion? OTOH, the `address` field will be defaulted by k0s's address discovery if omitted. There wouldn't be an easy way to let users use the auto-detected address as the bind address, then.

Anyway, I'm fine with the `onlyBindToAddress` logic as well, but wanted to at least ask about it.

Comment (twz123) [inline docs/configuration.md:120]: An alternative name could be `bindToAllInterfaces` which defaults to `true`. This would have to be implemented as a bool pointer though. `onlyBindToAddress` is a bit harder for me to understand.

**Review by twz123 — COMMENTED** [review]
Thanks for your dedication and perseverance, and sorry for the very late feedback. Looks good. Just some small fixes to use `net.JoinHostPort`. I'd like to get some quick feedback from other maintainers on the name of the config knob, 👍 otherwise.

File: docs/configuration.md
Comment (jnummelin) [reply docs/configuration.md:120]: IMO `bindAddress` makes more sense and is probably more common in general. At least I remember seeing such knobs on other things quite often. 😄

Comment (twz123) [reply docs/configuration.md:120]: Agreed. However, what about

> There wouldn't be an easy way to let users use the auto-detected address as the bind address, then.

Comment (jnummelin) [reply docs/configuration.md:120]: hmm, what's wrong in the current default behaviour of autodetecting the publish address and binding to all interfaces?

File: docs/configuration.md
Comment (pschichtel) [reply docs/configuration.md:58]: right

File: docs/configuration.md
Comment (pschichtel) [reply docs/configuration.md:120]: I read in other PRs that you generally prefer defaulting to false, so I'd assume other booleans are also like that. Not sure this is worth breaking the consistency, especially since this is more of a niche option.

Comment (pschichtel) [reply docs/configuration.md:120]: > IMO bindAddress makes more sense and is probably more common in general. At least I remember seeing such knobs on other things quite often.

bind-address is a common term yes, but rarely any software asks you to correctly configure several addresses for different purposes, that may or may not interact with each other. That's why I intended to reduce the bind-address parameter to what makes sense. I think the current documentation of `address` is misleading, given nothing actually binds to that address, it's just the internal address published to other components.

> hmm, what's wrong in the current default behaviour of autodetecting the publish address and binding to all interfaces?

The original issue #957 describe where this PR comes from: Multi-homed kubernetes hosts. In my case specially, my k0s hosts was also my edge router, so one of the interfaces the API server bound to was my internet uplink. With proper firewall rules in place no big deal, but it carries some risk. My setup since changed (for the better).

> There wouldn't be an easy way to let users use the auto-detected address as the bind address, then.

if the user configured `bindToAddressOnly: true` without actually configuring the `address`, then it would just use the auto-detected address, right? (probably worth documenting too)

As I said in https://github.com/k0sproject/k0s/pull/3824#discussion_r1622862280, `address` and `bindAddress` are strongly related. Configuring them incorrectly will break the cluster. You can't bind to an address that's not internally announced (unless NAT between components is supported, which would be ridiculous), so you have 2 options: bind to all interfaces and announce one of the addresses or bind to the address you announce. I can't see any other valid configuration (again: assuming NAT is not a thing here). 

Also... I'd assume (I don't have numbers on it) fairly few people actually have a need for this. Few kubernetes hosts are multi-homed and of those most are fine with binding all interfaces.

File: cmd/controller/certificates.go
Comment (pschichtel) [reply cmd/controller/certificates.go]: makes sense, done

File: docs/configuration.md
Comment (pschichtel) [reply docs/configuration.md]: I improved it I think, please have a look.

File: pkg/apis/k0s/v1beta1/api.go
Comment (jnummelin) [inline pkg/apis/k0s/v1beta1/api.go:42]:
```suggestion
	OnlyBindToAddress bool `json:"onlyBindToAddress,omitempty"`
```

IMO worth to keep the variable name and json key in sync

Comment (pschichtel) [reply pkg/apis/k0s/v1beta1/api.go:42]: ops, yeah I'll fix that in the evening.

**Comment by pschichtel** [comment]
Is that failing job just flakyness? It seems to exceeded a deadline unrelated to my change, right?

**Comment by twz123** [comment]
Yep, the inttests are quite a bit flaky, unfortunately.

File: docs/configuration.md
Comment (kke) [reply docs/configuration.md]: `bindExclusivelyToAddress`, `bindExclusive`, `bind(To)Address(Only)`, `addressBind`?

Somehow `onlyBindToAddress` sounds clunky to me.

Of course `address` (advertise) + `bindAddress: <address>` (bind) would be the best, I think there's an issue for that (#1880).

Comment (pschichtel) [reply docs/configuration.md]: > Somehow onlyBindToAddress sounds clunky to me.

I agree, I'm not super happy with the name, but at least it's descriptive.

> Of course address (advertise) + bindAddress: <address> (bind) would be the best

I actually don't agree here. I think `bindAddress` is not an option that should be offered, because the additional flexibility brings no benefit, only the risk that people misconfigure it.

**Review by twz123 — APPROVED** [review]
Thank you!

**Comment by mback2k** [comment]
`k0sctl` will need to be tweaked to support this as it currently tries to validate the API server coming up this way:
```
executing `curl -kso /dev/null --connect-timeout 20 -w \"%{http_code}\" \"https://localhost:6443/version\"`"
```
and then fails this way:
```
INFO ==> Running phase: Upgrade controllers 
INFO [ssh] ...: starting upgrade   
INFO [ssh] ...: waiting for the k0s service to start 
INFO * Running clean-up for phase: Acquire exclusive host lock 
INFO * Running clean-up for phase: Upgrade controllers 
INFO ==> Apply failed 
```

**Comment by twz123** [comment]
@mback2k Mind creating a [k0sctl issue](https://github.com/k0sproject/k0sctl/issues/new) for that?

**Pull request merged**
